keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,366 @@
+
+### Can we integrate with podresource API?
+
+|           | pod resource API | pod info API |
+| ----------| ---------------- | ------------ |
+| What information is returned with API | Container resources (devices, cpu_ids, memory, dynamic resources) | for Static and Regular Pods pod readiness Conditions and UID  |
+| Who is the owner of the data returned by API | kubelet, those data are not reported to Control Plane | static pods - kubelet, regular pods - kubelet, data are reported to kube-API |
+| What is the alternative for the API | reading the state files resource manager use | for regular pods - fetching pod's data with kube-API, for static pods - none |
+| Control Plane dependency | no | no |
+| Authorization | none, using unix socket to limit the clients to local workloads | none, using unix socket to limit the clients to local workloads |
+| access | Read-Only | Read-Only|
+| clients type / use cases | Device Plugins, CSI, Topology aware scheduling | CNI, monitoring and observability tools |
+
+Podresource API looks like a good fit for this endpoint but it looks like the use cases and clients type might not overlap. Our major concern is how will this integration impact the shape of this API in the future.
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads. If kubelet is in the init phase and not all pod’s information is known, then the API should report that information. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but kubelet is working properly this API should return most recent data about local pods that are computed by kubelet even if those data were not reported or accepted by Kube-API. |
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system integrity. However, already running pods should not be dependent on Control Plane availability or connectivity issues. When Node is healthy but has problems accessing Kube-API running pods, statuses should be refreshed and stored as usual.
```

Comment:
kube-apiserver not accessible from a given node isn't the sign of its unavailability. It's possible that the node lost its network connectivity and the rest of the system will eventually consider the node down. So this assumptions is only partially valid, in very narrow case, which I'd like to be explicitly laid out here.

Reply:
Fair point, added some additional content to clarify this. A high level summary is that since this would only impact node-local components, and the Node would essentially be cut off from the rest of the cluster in this state, it should not be a concern for this API.